### ISSUE TITLE
Infinite retry loop can be seen as spam attack by relays

### ISSUE THREAD

**Issue by alltheseas** [Yonle/bostr#12]
From relay operator @mikedilger

>  have had to block bostr on my relay due to an instant-retry-in-an-infinite-loop-that-is-essentially-a-dos-attack issue. In fact I think I'm blocking half of Japan at this point just because of this "bug".

**Comment by Yonle** [comment]
Yes. I have been expected that this is could happening anytime soon. However resolving it may take a time to think for a solution.

Simply stop reconnecting after 3-7 attempts (or so) is brutal for a bouncer. But making a handler to wait other session could connect back to the upstream could seem fix the issue.

However just thinking how to put that in a code.

**Comment by Yonle** [comment]
Some special note, It was supposed to be used for personal / private use. Making it for public of course will harm the upstream.

**Comment by Yonle** [comment]
And another note, I've put 5 seconds of wait before reconnect since the very beginning.

**Comment by mikedilger** [comment]
> Simply stop reconnecting after 3-7 attempts (or so) is brutal for a bouncer.

I wouldn't stop trying, I would just slow down.  A 5-minute timeout would be plenty.  If you tried every 5 minutes for eternity I would be ok with that.

> I've put 5 seconds of wait before reconnect since the very beginning.

I'm getting about 27 retries per second per IP address, and there seem to be more and more IP addresses using bostr, so that's something like 40*27 connections per second, each one needing to setup IP, TCP, SSL, HTTP and Websockets, send in a first message. Then it gets closed from my side.  If I didn't close them my website would be essentially down, because that is how I discovered this. Eventually I figured out that blocking the IP addresses, not allowing the connections to establish in the first place, greatly relieved the strain and my website started working again.  But I don't want to block Japan. I love Japan. I want people in Japan to see my nostr notes.  I really just wish this could be fixed.

Here are my suggestions.  I am not claiming that these are not already in place, maybe some of them already are.  I didn't look at your code.

1. If the connection drops, wait for a timeout before trying again. 30 seconds would be a huge relief, but I'd prefer 5 minutes.  And if you can detect that the connection dropped in a way that indicates the server doesn't want you to retry (look at the HTTP status code, something like FORBIDDEN would be an indication that retrying won't work), then don't retry, store some kind of flag in the relay record that the relay has rejected the client (I'm not doing that, but someone might and it should be honored).  If you can't detect the server's intent (e.g. if SSL fails or IP drops without explanation), then retrying every 5 minutes is fine.
2. Timeout if the subscription drops before trying that same subscription again. 30 seconds would be okay at this layer I think.
3. Look at the CLOSED response if a relay CLOSED your subscription, and if it is catastrophic (blocked, invalid or error) then don't try that subscription again, instead mark the relay as unable to handle that kind of subscription. If the reason is 'rate-limited' then slow down somehow.  I would store a 'rate' field in the relay record.
4. Try not to open more than about 10 subscriptions. I know that's hard to get below 10 so I allow up to 24. I think that should be enough.

Maybe I'll take some time out to look through the code. I might be able to submit a PR.

**Comment by Yonle** [comment]
> 1. If the connection drops, wait for a timeout before trying again. 30 seconds would be a huge relief, but I'd prefer 5 minutes.  And if you can detect that the connection dropped in a way that indicates the server doesn't want you to retry (look at the HTTP status code, something like FORBIDDEN would be an indication that retrying won't work), then don't retry, store some kind of flag in the relay record that the relay has rejected the client (I'm not doing that, but someone might and it should be honored).  If you can't detect the server's intent (e.g. if SSL fails or IP drops without explanation), then retrying every 5 minutes is fine.

This could be fine for me. 

> 2. Timeout if the subscription drops before trying that same subscription again. 30 seconds would be okay at this layer I think.

It never tries to reuse subscription. In fact it does not handle CLOSED yet as majority of relays did not implemented it.

> 3. Look at the CLOSED response if a relay CLOSED your subscription, and if it is catastrophic (blocked, invalid or error) then don't try that subscription again, instead mark the relay as unable to handle that kind of subscription. If the reason is 'rate-limited' then slow down somehow.  I would store a 'rate' field in the relay record.

When you get `NOTICE` with the `rate-limited` reason, It could be either both publishing and requesting was being ratelimited, which ended up giving the client a small slowdown & confusion.

> 4. Try not to open more than about 10 subscriptions. I know that's hard to get below 10 so I allow up to 24. I think that should be enough.

You better tell to these nostr clients dev instead of me. Different relays often has different limit. So setting 10 at default is absurd. But for some reasons, I will implement this to be configureable.

**Comment by Yonle** [comment]
Configureable ratelimits has been implemented as per a3b4c20

For bostr admins, Please check config.js.example

**Comment by Yonle** [comment]
# Notice to all relays operator

Bostr will now put it is version on it is user agent. Things about ratelimit has been fixed since v2.0.4-dev.

If there are another bostr function that acts up even at this version, It could be due to unconfigured ratelimit options.
